convert-hf-to-gguf.py

```diff
@@ -91,10 +132,39 @@ def __init__(self, dir_model: Path, ftype: gguf.LlamaFileType, fname_out: Path,
             else:
                 logger.info(f"choosing --outtype bf16 from first tensor type ({first_tensor.dtype})")
                 self.ftype = gguf.LlamaFileType.MOSTLY_BF16
-        ftype_up: str = self.ftype.name.partition("_")[2].upper()
-        ftype_lw: str = ftype_up.lower()
-        # allow templating the file name with the output ftype, useful with the "auto" ftype
-        self.fname_out = fname_out.parent / fname_out.name.format(ftype_lw, outtype=ftype_lw, ftype=ftype_lw, OUTTYPE=ftype_up, FTYPE=ftype_up)
+
+        # Set model name based on latest metadata either provided or calculated from environment
+        def get_model_name(metadata, hyperparameters, dir_model, model_arch):
+            if metadata is not None and metadata.name is not None:
+                # Explicit Metadata Was Provided By User
+                return metadata.name
+            elif hyperparameters is not None and "model_type" in hyperparameters:
+                # Hugging Face Hyperparameter Model Type Already Provided
+                return hyperparameters["model_type"]
+            elif dir_model is not None and dir_model.name is not None:
+                # Use directory folder name
+                return dir_model.name
+            else:
+                return gguf.MODEL_ARCH_NAMES[model_arch]
+        self.model_name = get_model_name(self.metadata, self.hparams, self.dir_model, self.model_arch)
+
+        # Generate default filename based on model specification and available metadata
+        def extract_encoding_scheme(ftype):
+            # Extracts and converts the encoding scheme from the given file type name.
+            # e.g. 'gguf.LlamaFileType.ALL_F32' --> 'F32'
+            return ftype.name.partition("_")[2].upper()
+        expert_count = self.hparams["num_local_experts"] if "num_local_experts" in self.hparams else None
+        self.fname_default = f"{gguf.naming_convention(self.model_name, self.metadata.version, expert_count, self.parameter_count(), extract_encoding_scheme(self.ftype))}"
+
+        # Filename Output
+        if fname_out is not None:
+            # custom defined filename and path was provided
+            self.fname_out = fname_out
```

Comment:
Custom names should support type templates as before.

https://github.com/ggerganov/llama.cpp/blob/0df0aa8e43c3378975269a51f9b876c8692e70da/convert-hf-to-gguf.py#L94-L97

Otherwise, now when running

```console
$ python3 convert-hf-to-gguf.py --outfile some-model-name.{ftype}.gguf --outtype auto /some/model/dir/
```

then the output file name would be `some-model-name.{ftype}.gguf` instead of what it currently does on `master` (let's say the model has `bf16` tensors):
`some-model-name.bf16.gguf`.